src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs

```diff
@@ -214,6 +214,9 @@ private static ExternalProjectReference GetExternalProject(PackageSpec rootProje
             // Add project references
             request.ExternalProjects = projectReferenceClosure.ToList();
 
+            //Update the RestoreRequest.Project.RestoreMetadata.Sources to get the attributes from settings(e.g. AllowInsecureConnections).
+            request.Project.RestoreMetadata.Sources = sources.Select(s => s.PackageSource).ToList();
```

Comment:
This was my bad @heng-liu, I should've done this investigation the first time I provided feedback. 

I think we need a slight change here. 
I *think* my original implementation that I added here: https://github.com/NuGet/NuGet.Client/commit/4d2c5a4ca8e38aa62d3c1d3357076c2a3f701c9c is not ideal. 

Some background: 

- PackageSpec, or rather `Project` is considered to be the `model` for restore, the input. This will be created differently on the commandline (static graph vs standard) vs Visual Studio. It has a sources property, which is the processed sources based on the project path, `RestoreSources` msbuild property etc. It's the actual end sources. The general expectation is that the PackageSpec *does* not get modified. It is basically what is used for no-op.
- RestoreRequest has a list of `DependencyProviders`. 
These dependency providers are "shared" across projects, created within this classl,https://github.com/NuGet/NuGet.Client/blob/41e98d33e40293d8276015d7a01e4d90610fa9d8/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs#L175-L182. It basically ensures that restore for every project shares the same providers, the same SourceRepository and like resources. These dependency provides also happen to have repositories, and PackageSource that *should* match the ones from the settings file. Hence the whole GetEffectiveSources thing. 

When I added the original implementation, I used `_request.Project.RestoreMetadata.Sources`. I think that's incorrect. Those are *not* how the sources are used, they are used through `request.DependencyProviders` which is effectively *based* on the PackageSource list *from* the package spec. 

I should've used that and it's already used in a bunch of places, like for example, detecting http sources for telemetry, https://github.com/NuGet/NuGet.Client/blob/41e98d33e40293d8276015d7a01e4d90610fa9d8/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L155. 

Basically, PackageSpec on the request is expected to be immutable, see: https://github.com/NuGet/Home/issues/9041#issuecomment-1703410411. We basically clone it a bunch of time because of fear of these types of changes. 

Now this is a thing you've never caught yourself, and it's not *incorrect* to do this, but I think it'll be a smaller change if you just change it to use the dependencyproviders. 

You can try that change and see how it goes, tests might break. Like, not sure if everything is mocked 100%. If that becomes the case, please lmk and I can help or just do the change from using the package spec metadata to the dependencyprovider once, since I caused the pain anyways :)

Reply:
I think the review has been re-requested mistakenly,

Reply:
Thanks @nkolev92 for the detailed contexts! (I was wondering why the tests for no-op failed previously.)
I changed to use RemoteProviders in a testing branch and the [build ](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=8355913&view=results) shows the newly changed test failed. There is no warning in result.LockFile.LogMessages so it failed at [this line](https://github.com/NuGet/NuGet.Client/blob/dev-hengliu-optoutHttpWarnings-packageReferrenceRestore-test4/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs#L4098)
However, I can see the warning from the console when running the commandline.
Do you have any suggestions about fixing this test? Thanks!

Reply:
https://github.com/NuGet/NuGet.Client/blob/29f292950dfa50aea030bd74c3277e983eff711e/test/TestUtilities/Test.Utility/Commands/ProjectTestHelpers.cs#L223 has a bug. It's using a hardcoded list where it's using the `PathContext` source instead of the ones from the package spec. It should be using the PackageSpec ones. I'll do some quick testing to see if I can have a fix.

Reply:
I'd add a new CreateRestoreRequestMethod, where the source is `spec.RestoreMetadata.Sources`. 

That makes the test pass. Other tests would fail if you make the change in project test helpers directly. 
I know how to fix the ones I wrote, but there's many other tests that have package specs that are incomplete.

Reply:
Actually if you replace that line with: 
```cs
            var sources = spec.RestoreMetadata.Sources.Any() ?
                spec.RestoreMetadata.Sources :
                new List<PackageSource> { new PackageSource(pathContext.PackageSource) };
```

All tests seem to pass.

Reply:
If you end up choosing that, please create an issue to fix the tests and assign to me. I think our test helpers code should use *as much* of the production code as possible. I'm hoping to do some changes in the ProjectTestHelpers that'll help us avoid these problems.

Reply:
I spent about an hour doing some refactoring: https://github.com/NuGet/NuGet.Client/tree/dev-nkolev92-cleanupTestHelpers.

Reply:
I have https://github.com/NuGet/NuGet.Client/pull/5407 out.

Reply:
Thanks @nkolev92 for your help! Please take another look.